python/cloudtik/providers/_private/huaweicloud/config.py

```diff
@@ -0,0 +1,811 @@
+import logging
+import time
+
+import requests
+from huaweicloudsdkecs.v2 import ListServersDetailsRequest
+from huaweicloudsdkeip.v2 import CreatePublicipBandwidthOption, \
+    CreatePublicipOption, CreatePublicipRequest, \
+    CreatePublicipRequestBody, DeletePublicipRequest, ListPublicipsRequest
+from huaweicloudsdknat.v2 import CreateNatGatewayOption, \
+    CreateNatGatewayRequest, \
+    CreateNatGatewayRequestBody, CreateNatGatewaySnatRuleOption, \
+    CreateNatGatewaySnatRuleRequest, \
+    CreateNatGatewaySnatRuleRequestOption, DeleteNatGatewayRequest, \
+    DeleteNatGatewaySnatRuleRequest, \
+    ListNatGatewaySnatRulesRequest, \
+    ListNatGatewaysRequest
+from huaweicloudsdkvpc.v2 import AcceptVpcPeeringRequest, \
+    CreateSecurityGroupOption, \
+    CreateSecurityGroupRequest, \
+    CreateSecurityGroupRequestBody, CreateSecurityGroupRuleOption, \
+    CreateSecurityGroupRuleRequest, \
+    CreateSecurityGroupRuleRequestBody, CreateSubnetOption, \
+    CreateSubnetRequest, \
+    CreateSubnetRequestBody, \
+    CreateVpcOption, \
+    CreateVpcPeeringOption, CreateVpcPeeringRequest, \
+    CreateVpcPeeringRequestBody, CreateVpcRequest, \
+    CreateVpcRequestBody, DeleteSecurityGroupRequest, \
+    DeleteSecurityGroupRuleRequest, DeleteSubnetRequest, \
+    DeleteVpcPeeringRequest, \
+    DeleteVpcRequest, ListRouteTablesRequest, \
+    ListSecurityGroupRulesRequest, ListSecurityGroupsRequest, \
+    ListSubnetsRequest, \
+    ListVpcPeeringsRequest, ListVpcsRequest, RouteTableRoute, ShowVpcRequest, \
+    UpdateRouteTableReq, UpdateRoutetableReqBody, UpdateRouteTableRequest, \
+    VpcInfo
+
+from cloudtik.core._private.cli_logger import cf, cli_logger
+from cloudtik.core._private.utils import check_cidr_conflict, \
+    is_managed_cloud_storage, \
+    is_peering_firewall_allow_ssh_only, \
+    is_peering_firewall_allow_working_subnet, is_use_peering_vpc, \
+    is_use_working_vpc
+from cloudtik.core.workspace_provider import Existence
+from cloudtik.providers._private.huaweicloud.utils import make_ecs_client, \
+    make_eip_client, make_nat_client, make_vpc_client
+
+logger = logging.getLogger(__name__)
+
+HWC_WORKSPACE_NUM_CREATION_STEPS = 6
+HWC_WORKSPACE_NUM_DELETION_STEPS = 6
+HWC_WORKSPACE_TARGET_RESOURCES = 5
+HWC_RESOURCE_NAME_PREFIX = 'cloudtik-{}'
+HWC_WORKSPACE_VPC_NAME = HWC_RESOURCE_NAME_PREFIX + '-vpc'
+HWC_WORKSPACE_DEFAULT_CIDR_PREFIX = '192.168.'
+HWC_WORKSPACE_VPC_DEFAULT_CIDR = HWC_WORKSPACE_DEFAULT_CIDR_PREFIX + '0.0/16'
+HWC_WORKSPACE_SUBNET_NAME = HWC_RESOURCE_NAME_PREFIX + '-{}-subnet'
+HWC_WORKSPACE_NAT_NAME = HWC_RESOURCE_NAME_PREFIX + '-nat'
+HWC_WORKSPACE_SG_NAME = HWC_RESOURCE_NAME_PREFIX + '-sg'
+HWC_WORKSPACE_VPC_PEERING_NAME = HWC_RESOURCE_NAME_PREFIX + '-peering'
+HWC_WORKSPACE_EIP_BANDWIDTH_NAME = HWC_RESOURCE_NAME_PREFIX + '-bandwidth'
+HWC_WORKSPACE_VPC_SUBNETS_COUNT = 2
+HWC_VM_METADATA_URL = 'http://169.254.169.254/latest/metadata/'
+
+
+def create_huaweicloud_workspace(config):
```

Comment:
CloudTik workspace includes three aspects of cloud resources: 1. networking, 2. cloud storage and 3. instance credentials/identities (with RBAC).
Networking part is implemented. cloud storage is marked TODO. 

Using the basic credentials, everything should be working. The design idea behind instance credentials/identities (with RBAC) is we should avoid using and passing secrets based credentials to cluster node. And the cluster node use instance credentials which give the access to cloud resources. 

head node and worker node will have different instance credential because head node instance credential needs assign the role of creating instances and accessing the cloud storage. While worker node don't need the permission of creating instances, it has only the need for accessing the cloud storage.

Of course, you can use another PR to implement instance based credential separately.

Reply:
Thank you for these clarification, I check HuaweiCloud document again, it have support the three features you mentioned, but have different name, I will implement in next commit:
1. networking == HuweiCLoud VPC
2. cloud storage == HuaweiCloud OBS
3. instance credentials == HuaweiCloud [IAM agency and instance metadata temporary AK/SK](1)

[1]: https://support.huaweicloud.com/intl/en-us/bestpractice-iam/iam_0511.html